File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/SkyMallOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.ConfigAccess;
import io.github.chindeaone.collectiontracker.config.core.Position;
import io.github.chindeaone.collectiontracker.util.HypixelUtils;
import io.github.chindeaone.collectiontracker.util.chat.ChatListener;
import io.github.chindeaone.collectiontracker.util.rendering.RenderUtils;
import io.github.chindeaone.collectiontracker.util.rendering.TextUtils;
import io.github.chindeaone.collectiontracker.util.tab.MiningStatsWidget;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SkyMallOverlay implements AbstractOverlay {

    private final Position position = ConfigAccess.getSkyMallPosition();
    private final List<String> skyMallOverlayLines = new ArrayList<>();
    private boolean renderingAllowed  = true;

    @Override
    public String overlayLabel() {
        return "SkyMall Overlay";
    }

    @Override
    public Position position() {
        return this.position;
    }

    @Override
    public boolean isEnabled() {
        return ConfigAccess.isSkyMallEnabled() && HypixelUtils.isOnSkyblock();
    }

    @Override
    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    @Override
    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    @Override
    public void render(GuiGraphics context) {
        if (!isEnabled()) return;
        List<String> lines = getSkyMallLines();

        if (lines.isEmpty()) return;

        RenderUtils.drawOverlayFrame(context, position, () ->
                RenderUtils.renderStrings(context, lines)
        );
    }

    @Override
    public void updateDimensions() {
        if (!isEnabled()) return;
        List<String> lines = getSkyMallLines();
        if (lines.isEmpty()) return;

        Font fr = Minecraft.getInstance().font;
        int maxW = 0;
        for (String l : lines) maxW = Math.max(maxW, fr.width(l));
        int h = fr.lineHeight * lines.size();

        position.setDimensions(maxW, h);
    }

    private List<String> getSkyMallLines() {
        skyMallOverlayLines.clear();
        if (ConfigAccess.isSkyMallInMiningIslandsOnly() && MiningStatsWidget.getCurrentMiningIsland() == null) return Collections.emptyList();

        skyMallOverlayLines.add("§bSky Mall: " + ChatListener.getCurrentSkyMallBuff());
        skyMallOverlayLines.add(TextUtils.updateTimer());
        return  skyMallOverlayLines;
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/SkyMallOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.config.enableSkyMall
import io.github.chindeaone.collectiontracker.config.skyMallInMiningIslandsOnly
import io.github.chindeaone.collectiontracker.config.skyMallPosition
import io.github.chindeaone.collectiontracker.utils.chat.ChatListener.currentSkyMallBuff
import io.github.chindeaone.collectiontracker.utils.world.IslandTracker.currentMiningIsland
import io.github.chindeaone.collectiontracker.utils.world.MiningMapping.miningIslands

class SkyMallOverlay : AbstractRotatingPerksOverlay() {
    override val overlayLabel: String = "Sky Mall"

    override val position: Position get() = skyMallPosition

    override val isEnabled: Boolean get() = enableSkyMall

    override val buffPrefix get() = "§bSky Mall"

    override val currentBuff: String get() = currentSkyMallBuff

    override val isIslandAllowed: Boolean get() = !skyMallInMiningIslandsOnly || miningIslands.contains(currentMiningIsland)
}